rfd/0125-dynamic-auto-discovery-config.md

```diff
@@ -0,0 +1,406 @@
+}
+
+// GCPMatcher matches GCP resources.
+message GCPMatcher {
+	// Types are GKE resource types to match: "gke".
+	repeated string Types = 1;
+	// Locations are GKE locations to search resources for.
+	repeated string Locations = 2;
+	// Tags are GCP labels to match.
+  wrappers.LabelValues Tags = 3;
+	// ProjectIDs are the GCP project ID where the resources are deployed.
+	repeated string ProjectIDs = 4;
+}
+
+```
+
+### Backward Compatibility
+
+#### `discovery_service.selector`
+If the config has this field but Teleport version is no aware of that, the user must remove it before they can start the Discovery Service.
+
+#### Watcher on `DiscoveryConfig`
+If the `discovery_service` tries to set up a watcher over `DiscoveryConfig` but Teleport Auth/Proxy is not yet aware of this resource, then `discovery_service` will log a warning but continue with the static configuration.
+
+### UX
+
+#### `discovery_service` new service configuration properties
+Users need to add the `selector` property and then a list of label matchers.
+
+Example:
+```yaml
+discovery_service:
+  enabled: "yes"
+  selector:
+    labelA: value1 # matches against a single value
+    labelB: [value2, value3] # matches against multiple values
+    labelC: "*" # matches on all values
+```
+
+#### WebAPI: manage `DiscoveryConfig` resource
+The following endpoints must be create to be used by WebAPI:
+```
+Methods:
+GET .../discoveryconfig
+GET .../discoveryconfig/:id
+POST .../discoveryconfig
+PUT .../discoveryconfig/:id
+DELETE .../discoveryconfig/:id
+```
+
+JSON representation:
+```json
+{
+   "aws":[
+      {
+         "types":[
+            "ec2"
+         ],
+         "regions":[
+            "us-east-1",
+            "us-west-1"
+         ],
+         "tags":{
+            "*":"*"
+         },
+         "install":{
+            "join_params":{
+               "token_name":"aws-discovery-iam-token"
+            },
+            "script_name":"default-installer"
+         },
+         "ssm":{
+            "document_name":"TeleportDiscoveryInstaller"
+         }
+      }
+   ],
+   "azure":[
+      {
+         "types":[
+            "aks"
+         ],
+         "regions":[
+            "eastus",
+            "westus"
+         ],
+         "subscriptions":[
+            "11111111-2222-3333-4444-555555555555"
+         ],
+         "resource_groups":[
+            "group1",
+            "group2"
+         ],
+         "tags":{
+            "*":"*"
+         }
+      }
+   ],
+   "gcp":[
+      {
+         "types":[
+            "gke"
+         ],
+         "locations":[
+            "*"
+         ],
+         "tags":{
+            "*":"*"
+         },
+         "project_ids":[
+            "myproject"
+         ]
+      }
+   ]
+}
+```
+
+#### `tctl`: manage `DiscoveryConfig` resource
+Users must be able to create, list, read, update and delete `DiscoveryConfig`s using `tctl`.
+
+Example of `tctl get discovery_config`:
+```
+$ tctl get discovery_config
+kind: DiscoveryConfig
```

Comment:
Out of context: 
@r0mant @xinding33 : I would love to see Dynamic Configuration resource support in our web UX. Do you know if we have any plan to do it in the feature ?

Reply:
The ultimate goal is for users to be able to deploy a DiscoveryService/agent and configure its matchers dynamically through the Web UX.

We didn't include everything in the RFD to ensure we create a tool/primitive that later we can re-use for any other purpose.
We could change the RFD to describe the entire flow but I think that would be too big and could lead to primitives that are too focused on achieving a certain goal instead of a more generic one (without being too generic, ofc).